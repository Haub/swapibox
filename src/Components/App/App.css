* {
  margin: 0 auto;
  width: 100%;
  font-family: lora;
  text-align: center; }

.header {
  height: 150px;
  padding: 20px; }

.title {
  font-size: 4rem;
  text-shadow: 1px 1px #523F38; }
