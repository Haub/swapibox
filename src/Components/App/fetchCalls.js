export const fetchData = async (value) => {
    try {
      const url = `https://swapi.co/api/${value}/`
      const response = await fetch(url)
      const data = await response.json()
	  return data
    } catch(error) {
      console.log(error.message)
    }
  }

export const fetchHome = async (url) => {
  try{
    const response = await fetch(url)
    const homeData = await response.json()
    const homeObj = { homeworld: homeData.name, population: homeData.population }
    return homeObj
  } catch(error) {
    console.log(error.message)
  }
}

export const fetchSpecies = async (url) => {
  try{
    const response = await fetch(url)
    const speciesData = await response.json()
    return speciesData.name
  } catch(error) {
    console.log(error.message)
  }
}

export const fetchPlanets = async (url) => {
  try {
    const response = await fetch(url)
    const planetData = await response.json()
    console.log(planetData)
     
    // console.log(planetObj)
    return planetObj
  } catch(error) {
    console.log(error.message)
  }
}