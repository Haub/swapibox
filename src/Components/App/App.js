import React, { Component } from 'react';
import './App.css';
import Landing from '../Landing/Landing';
import Button from '../Buttons/Button';
import { CategoryContainer } from '../CategoryContainer/CategoryContainer';
import { fetchData, fetchHome, fetchSpecies, fetchResidents } from '../App/fetchCalls.js'

class App extends Component {
	constructor() {
		super();

		this.state = {
      displayedCards: [],
      currentlyDisplayed: '',
      people: [],
      planets: [],
      vehicles: [],
      favorites: [],
      buttons: {people: false,
                planets: false,
                vehicles: false,
                favorites:false,
      },
      landing: true
		}
	}

  populateCards = async (event) => {
    let getData;
    const {value} = event.target;
    if(`!this.state.${value}.length`) {
      const data = await fetchData(value)
        if(value === 'people') {
          getData = await this.filterPeople(data.results)
        }
        if(value === 'planets') {
          getData = await this.filterPlanets(data.results)
        }
        if(value === 'vehicles') {
          getData = await this.filterVehicles(data.results)
        }
      this.setState({
        [value]: getData,
        buttons: {[value]: true},
      })
    }
      this.setState({
        displayedCards: this.state[value],
        currentlyDisplayed: value
      })
  }

  filterPeople = (cardItems) => {
    const character = cardItems.map(async card => {
      const name = card.name
      const homeSearch = await fetchHome(card.homeworld)
      const species = await fetchSpecies(card.species)
      const homeworld = homeSearch.homeworld
      const population = homeSearch.population
      const personObj = await {name, species, homeworld, population}
      return personObj
    })
    return Promise.all(character)
  }

  filterPlanets = (cardItems) => {
    const planet = cardItems.map(async card => {
      const name = card.name
      const terrain = card.terrain
      const population = card.population
      const climate = card.climate 
      const residents = await fetchResidents(card.residents)
      const planetObj = await {name, terrain, population, climate, residents}
      return planetObj
    })
    return Promise.all(planet)
  }

  filterVehicles = (cardItems) => {
    const vehicle = cardItems.map(async card => {
      const name = card.name
      const model = card.model
      const vehicle_class = card.vehicle_class
      const passengers = card.passengers
      const vehicleObj = {name, model, vehicle_class, passengers}
      return vehicleObj
    })
    return Promise.all(vehicle)
  }
  
  removeLanding = () => {
    this.setState({
      landing: false
    })
  }

  render() {
    return(
      <div>
        <header className="header">
          <h1 className="title">SWAPIbox</h1>
          <div className="button-box">
            <Button 
              value='people' 
              populateCards={this.populateCards}
            />
            <Button 
              value='planets'
              populateCards={this.populateCards}
            />
            <Button 
              value='vehicles' 
              populateCards={this.populateCards}
            />
            <Button 
              value='favorites'
              // populateCards={this.populateCards}
            />
          </div>
        </header>
        <CategoryContainer stateArray={this.state.displayedCards}
                            currentlyDisplayed={this.state.currentlyDisplayed}/>
        {this.state.landing &&
        <Landing removeLanding={this.removeLanding}/>}
      </div>
    )
  }
}

export default App;
