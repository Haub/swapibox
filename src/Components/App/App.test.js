import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow } from 'enzyme';

describe('App', () => {
  let wrapper;
  let expected;

  beforeEach(() => {
    wrapper = shallow(<App />);
  });

  it('matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should invoke populateCards when a button is pressed', () => {
    let mockPopulate = jest.fn();
    wrapper.find('.button-box').simulate('click');
    expect().toHaveBeenCalled();
  });

  it('should render the category container with the correct props', () => {
    expect(wrapper.find(CategoryContainer).prop('stateArray')).toEqual([]);
    expect(wrapper.find(CategoryContainer).prop('addToFavorites')).toEqual(wrapper.instance().addToFavorites);
  });

  it('should update state when populateCards is invoked', () => {
    const initialState = [];
    const mockCategory = {
      homeworld: "Tatooine",
      id: 1135694282614.0696,
      name: "Luke Skywalker",
      population: "200000",
      species: "Human",
      type: "people" };
    expected = [{
      homeworld: "Tatooine",
      id: 1135694282614.0696,
      name: "Luke Skywalker",
      population: "200000",
      species: "Human",
      type: "people" }];
    // Date.now() * Math.random() = jest.fn().mockImplementation(() => 1135694282614.0696);
    wrapper.setState({ displayedCards: initialState });
    wrapper.instance().populateCards(mockCategory);
    expect(wrapper.state('displayedCards')).toEqual(expected);
    expect(wrapper.state('displayedCards').length).toEqual(1);
  });

  it('should return an array of people when filterPeople is invoked', () => {
    mockInput = [{
      name: "Luke Skywalker",
      species: "https://swapi.co/api/species/1/",
      homeworld: "https://swapi.co/api/planets/1/",
    }];
    expected = [{
      homeworld: "Tatooine",
      id: 1135694282614.0696,
      name: "Luke Skywalker",
      population: "200000",
      species: "Human",
      type: "people" }];
    expect.instance().filterPeople().toHaveBeenCalledWith(mockInput);
    expect(wrapper.state('people')).toEqual(expected)
  });

  it('should return an array of planets when filterPlanets is invoked', () => {
    mockInput = [{
      climate: "temperate",
      name: "Alderaan",
      population: "2000000000",
      residents: [
        "https://swapi.co/api/people/5/",
        "https://swapi.co/api/people/68/",
        "https://swapi.co/api/people/81/" ],
      terrain: "grasslands, mountains"
    }];
    expected = [{
      climate: "temperate",
      id: 1410520490424.7056,
      name: "Alderaan",
      population: "2000000000",
      residents: [
        "Leia Organa",
        "Bail Prestor Organa",
        "Raymus Antilles" ],
      terrain: "grasslands, mountains",
      type: "planets" }];
    expect.instance().filterPlanets().toHaveBeenCalledWith(mockInput);
    expect(wrapper.state('planets')).toEqual(expected)
  });

  it('should return an array of vehicles when filterVehicles is invoked', () => {
    mockInput = [{
      name: "Sand Crawler",
      passengers: "30",
      vehicle_class: "wheeled",
      model: "Digger Crawler",
    }];
    expected = [{
      id: 105989059159.61761,
      model: "Digger Crawler",
      name: "Sand Crawler",
      passengers: "30",
      type: "vehicles",
      vehicleClass: "wheeled" }];
    expect.instance().filterVehicles().toHaveBeenCalledWith(mockInput);
    expect(wrapper.state('vehicles')).toEqual(expected);
  });

  it('should update state when removeLanding function is invoked', () => {
    const initialState = true;
    expected = false;
    wrapper.find('.fade').simulate('click');
    expect(removeLanding).toHaveBeenCalled();
    expect(wrapper.state('landing')).toEqual(expected);
  });

  it('should update state when addToFavorites is invoked', () => {
    const initialState = [];
    const mockFavorite = [{
      id: 105989059159.61761,
      model: "Digger Crawler",
      name: "Sand Crawler",
      passengers: "30",
      type: "vehicles",
      vehicleClass: "wheeled" }];

    expected = [{
      id: 105989059159.61761,
      model: "Digger Crawler",
      name: "Sand Crawler",
      passengers: "30",
      type: "vehicles",
      vehicleClass: "wheeled" }];
    wrapper.setState({ favorites: initialState });
    wrapper.instance().addToFavorites(mockFavorite);
    expect(wrapper.state('favorites')).toEqual(expected);
    expect(wrapper.state('favorites').length).toEqual(1);
  });

  it('should check the state of favorites for duplicates when checkDuplicates is invoked', () => {
    const mockParam = "Sand Crawler";
    const initialState = [{
      id: 105989059159.61761,
      model: "Digger Crawler",
      name: "Sand Crawler",
      passengers: "30",
      type: "vehicles",
      vehicleClass: "wheeled" }, {
      id: 105989059159.61761,
      model: "Digger Crawler",
      name: "Sand Crawler",
      passengers: "30",
      type: "vehicles",
      vehicleClass: "wheeled" }]];
    expected = [];
    wrapper.setState({ favorites: initialState });
    wrapper.instance().checkDuplicates(mockParam);
    expect(checkDuplicates).toHaveBeenCalledWith(mockParam);
    expect(wrapper.state(favorites)).toEqual(expected);
  });

  it('should invoke displayFavorites when a favorites button is pressed', () => {
    wrapper.find('.button-box').simulate('click');
    expect(displayFavorites).toHaveBeenCalled();
  });

  it('should update state when displayFavorites is invoked', () => {
    const initialState = [];
    expected = [{
      homeworld: "Tatooine",
      id: 1135694282614.0696,
      name: "Luke Skywalker",
      population: "200000",
      species: "Human",
      type: "people" }]
    wrapper.setState({ favorites: initialState });
    wrapper.instance().displayFavorites();
    expect(wrapper.state(displayedCards)).toEqual(expected);
  });
});





