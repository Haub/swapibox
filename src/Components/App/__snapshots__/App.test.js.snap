// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <header
          className="header"
>
          <h1
                    className="title"
          >
                    SWAPIbox
          </h1>
          <div
                    className="button-box"
          >
                    <Button
                              populateCards={[Function]}
                              value="people"
                    />
                    <Button
                              populateCards={[Function]}
                              value="planets"
                    />
                    <Button
                              populateCards={[Function]}
                              value="vehicles"
                    />
                    <Button
                              value="favorites"
                    />
          </div>
</header>,
        <Unknown
          currentlyDisplayed=""
          stateArray={Array []}
/>,
        <Landing
          removeLanding={[Function]}
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1
              className="title"
>
              SWAPIbox
</h1>,
            <div
              className="button-box"
>
              <Button
                            populateCards={[Function]}
                            value="people"
              />
              <Button
                            populateCards={[Function]}
                            value="planets"
              />
              <Button
                            populateCards={[Function]}
                            value="vehicles"
              />
              <Button
                            value="favorites"
              />
</div>,
          ],
          "className": "header",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "SWAPIbox",
              "className": "title",
            },
            "ref": null,
            "rendered": "SWAPIbox",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Button
                  populateCards={[Function]}
                  value="people"
/>,
                <Button
                  populateCards={[Function]}
                  value="planets"
/>,
                <Button
                  populateCards={[Function]}
                  value="vehicles"
/>,
                <Button
                  value="favorites"
/>,
              ],
              "className": "button-box",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "populateCards": [Function],
                  "value": "people",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "populateCards": [Function],
                  "value": "planets",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "populateCards": [Function],
                  "value": "vehicles",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "value": "favorites",
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": "div",
          },
        ],
        "type": "header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "currentlyDisplayed": "",
          "stateArray": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "removeLanding": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header
            className="header"
>
            <h1
                        className="title"
            >
                        SWAPIbox
            </h1>
            <div
                        className="button-box"
            >
                        <Button
                                    populateCards={[Function]}
                                    value="people"
                        />
                        <Button
                                    populateCards={[Function]}
                                    value="planets"
                        />
                        <Button
                                    populateCards={[Function]}
                                    value="vehicles"
                        />
                        <Button
                                    value="favorites"
                        />
            </div>
</header>,
          <Unknown
            currentlyDisplayed=""
            stateArray={Array []}
/>,
          <Landing
            removeLanding={[Function]}
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1
                className="title"
>
                SWAPIbox
</h1>,
              <div
                className="button-box"
>
                <Button
                                populateCards={[Function]}
                                value="people"
                />
                <Button
                                populateCards={[Function]}
                                value="planets"
                />
                <Button
                                populateCards={[Function]}
                                value="vehicles"
                />
                <Button
                                value="favorites"
                />
</div>,
            ],
            "className": "header",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "SWAPIbox",
                "className": "title",
              },
              "ref": null,
              "rendered": "SWAPIbox",
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Button
                    populateCards={[Function]}
                    value="people"
/>,
                  <Button
                    populateCards={[Function]}
                    value="planets"
/>,
                  <Button
                    populateCards={[Function]}
                    value="vehicles"
/>,
                  <Button
                    value="favorites"
/>,
                ],
                "className": "button-box",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "populateCards": [Function],
                    "value": "people",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "populateCards": [Function],
                    "value": "planets",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "populateCards": [Function],
                    "value": "vehicles",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "value": "favorites",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": "div",
            },
          ],
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "currentlyDisplayed": "",
            "stateArray": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "removeLanding": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
